File: go/problem_56.go
package main

func graphColorable(adj [][]int, k int) bool {
	n := len(adj)
	colors := make([]int, n)
	var bt func(node int) bool
	bt = func(node int) bool {
		if node == n {
			return true
		}
		for c := 1; c <= k; c++ {
			ok := true
			for nb := 0; nb < n; nb++ {
				if adj[node][nb] == 1 && colors[nb] == c {
					ok = false
					break
				}
			}
			if ok {
				colors[node] = c
				if bt(node + 1) {
					return true
				}
				colors[node] = 0
			}
		}
		return false
	}
	return bt(0)
}

func problem56() {
	println("Problem 56:")
	triangle := [][]int{{0, 1, 1}, {1, 0, 1}, {1, 1, 0}}
	if !graphColorable(triangle, 3) {
		panic("triangle needs 3 colors")
	}
	if graphColorable(triangle, 2) {
		panic("triangle cannot be 2-colored")
	}
	println("All Tests Passed")
}

File: python/problem_56.py
from typing import List


def graph_colorable(adj: List[List[int]], k: int) -> bool:
    n = len(adj)
    colors = [0] * n

    def bt(node: int) -> bool:
        if node == n:
            return True
        for c in range(1, k + 1):
            if all(adj[node][nb] == 0 or colors[nb] != c for nb in range(n)):
                colors[node] = c
                if bt(node + 1):
                    return True
                colors[node] = 0
        return False

    return bt(0)


def main():
    print("Problem 56:")
    triangle = [[0,1,1],[1,0,1],[1,1,0]]
    assert     graph_colorable(triangle, 3)
    assert not graph_colorable(triangle, 2)
    print("All Tests Passed")
